tutorial/cpp/4-GPU.cpp

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD+Patents license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <cstdio>
+#include <cstdlib>
+
+#include <faiss/IndexFlat.h>
+
+
+int main() {
+    int d = 64;                            // dimension
+    int nb = 100000;                       // database size
+    int nq = 10000;                        // nb of queries
+
+    float *xb = new float[d * nb];
+    float *xq = new float[d * nq];
+
+    for(int i = 0; i < nb; i++) {
+        for(int j = 0; j < d; j++)
+            xb[d * i + j] = drand48();
+        xb[d * i] += i / 1000.;
+    }
+
+    for(int i = 0; i < nq; i++) {
+        for(int j = 0; j < d; j++)
+            xq[d * i + j] = drand48();
+        xq[d * i] += i / 1000.;
+    }
+
+    faiss::gpu::GpuResources *res = new faiss::gpu::StandardGpuResources;
+
+    faiss::GpuIndexFlatL2 index(res, d);   // call constructor
+
+    printf("is_trained = %s\n", index.is_trained ? "true" : "false");
+    index.add(nb, xb);                     // add vectors to the index
+    printf("ntotal = %ld\n", index.ntotal);
+
+    int k = 4;
+
+    {       // sanity check: search 5 first vectors of xb
+        long *I = new long[k * 5];
```

Comment:
When relying on C++11, one can wrap these data allocations around a `unique_ptr`. Not only it would be automatically freed when falling out of scope, it would protect also prevent leaks in case of exceptions.

```c++
#include <memory>

// ...

auto I = std::unique_ptr<long[]>(new long[k * 5]);
auto D = std::unique_ptr<float[]>(new float[k * 5]);

index.search(5, xb, k, D.get(), I.get());
```